lxd/main_daemon.go

```diff
@@ -76,23 +76,42 @@ func (c *cmdDaemon) Run(cmd *cobra.Command, args []string) error {
 	chIgnore := make(chan os.Signal)
 	signal.Notify(chIgnore, unix.SIGHUP)
 
+	chForceShutdown := make(chan bool)
+	chDone := make(chan bool)
+
 	s := d.State()
 	select {
 	case sig := <-ch:
 		if sig == unix.SIGPWR {
-			logger.Infof("Received '%s signal', shutting down containers", sig)
+			logger.Infof("Received '%s signal', shutting down instances", sig)
 			d.Kill()
-			containersShutdown(s)
+			instancesShutdown(s)
 			networkShutdown(s)
 		} else {
 			logger.Infof("Received '%s signal', exiting", sig)
 			d.Kill()
 		}
 
 	case <-d.shutdownChan:
-		logger.Infof("Asked to shutdown by API, shutting down containers")
+		logger.Infof("Asked to shutdown by API, waiting for all operations to finish")
+		// Cancelling the context will make everyone aware that we're shutting down.
+		d.cancel()
+
+		go func() {
+			waitForOperations(s, chForceShutdown)
+			chDone <- true
```

Comment:
@monstermunchkin same with this channel, can you use empty struct and close when done, as otherwise using a bool type indicates that you could send false too, but there is no difference in meaning here.